### PR TITLE
Creating and example for Bayesian Neural Networks using TensorFlow Probability

### PR DESCRIPTION
**Pull request by ksalama**
@fchollet - Hello François!  Kindly review this example, which demonstrates how to build basic probabilistic Bayesian neural networks to account for data and model uncertainty, using TensorFlow Probability library.

### CODE REVIEW

**Comment by ksalama** [comment]
@fchollet - a kind reminder to take a look at this PR. Thanks :-)

File: examples/generative/bayesian_neural_networks_wine.py [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+        .map(lambda x, y: (x, tf.cast(y, tf.float32)))
+    )
+
+    train_dataset = (
+        dataset.take(train_size)
+        .shuffle(buffer_size=train_size)
+        .batch(batch_size)
+        .repeat(1)
+    )
+    test_dataset = dataset.skip(train_size).batch(batch_size).repeat(1)
```
Comment (fchollet) [inline examples/generative/bayesian_neural_networks_wine.py:83]: Not sure, but I think there's a chance that the test dataset overlaps with the training dataset. Because you have a `shuffle(dataset_size)` before `take`. The shuffle order will be different at each iteration (=epoch).

I recommend downloading the data from UCI to avoid these issues https://archive.ics.uci.edu/ml/datasets/wine+quality

File: examples/generative/bayesian_neural_networks_wine.py [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+def run_experiment(model, loss, train_dataset, test_dataset):
+
+    model.compile(
+        optimizer=keras.optimizers.RMSprop(learning_rate=learning_rate),
+        loss=loss,
+        metrics=[keras.metrics.RootMeanSquaredError()],
+    )
+
+    print("Start training the model...")
+    model.fit(train_dataset, epochs=num_epochs)
```
Comment (fchollet) [inline examples/generative/bayesian_neural_networks_wine.py:105]: Need a validation dataset to monitor during training.

File: examples/generative/bayesian_neural_networks_wine.py [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+    print("Model training finished.")
+    _, rmse = model.evaluate(train_dataset, verbose=0)
+    print(f"Train RMSE: {round(rmse, 3)}")
```
Comment (fchollet) [inline examples/generative/bayesian_neural_networks_wine.py:108]: What's interesting is validation error, training error will not be informative

File: examples/generative/bayesian_neural_networks_wine.py [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+"""
+## Experiment 1: standard neural network
+
+We create a standard deterministic neural network model as a baseline.
+"""
+
+
+def create_baseline_model():
+    inputs = create_model_inputs()
+    features = keras.layers.concatenate(list(inputs.values()))
```
Comment (fchollet) [inline examples/generative/bayesian_neural_networks_wine.py:152]: You should sort the features by name to avoid indeterminism

File: examples/generative/bayesian_neural_networks_wine.py [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+    features = layers.BatchNormalization()(features)
```
Comment (fchollet) [inline examples/generative/bayesian_neural_networks_wine.py:153]: BatchNormalization may not work well with a small batch size (the default in `get_train_and_test_splits` is 1). Consider using a Normalization layer instead (and call `adapt` on it with the concatenated features)

**Review by fchollet — COMMENTED** [review]
Thanks for the PR

File: examples/generative/bayesian_neural_networks_wine.py (hunk 2 above)
Comment (ksalama) [reply examples/generative/bayesian_neural_networks_wine.py:105]: Done

File: examples/generative/bayesian_neural_networks_wine.py (hunk 3 above)
Comment (ksalama) [reply examples/generative/bayesian_neural_networks_wine.py:108]: I am actually reporting both.  Do you recommend removing the train error?

File: examples/generative/bayesian_neural_networks_wine.py (hunk 4 above)
Comment (ksalama) [reply examples/generative/bayesian_neural_networks_wine.py:152]: Done

File: examples/generative/bayesian_neural_networks_wine.py (hunk 5 above)
Comment (ksalama) [reply examples/generative/bayesian_neural_networks_wine.py:153]: Done

File: examples/generative/bayesian_neural_networks_wine.py (hunk 1 above)
Comment (ksalama) [reply examples/generative/bayesian_neural_networks_wine.py:83]: Good point. I removed the first shuffle. I prefer to use the tfds to demonstrate the use of TensorFlow Datasets with Keras (if you are Ok with this).

**Comment by ksalama** [comment]
@fchollet - Thank you for your review. I have addressed your comments.

File: examples/generative/bayesian_neural_networks_wine.py [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+    dataset = tfds.load(name="wine_quality", as_supervised=True, split="train").map(
+        lambda x, y: (x, tf.cast(y, tf.float32))
+    )
+
+    train_dataset = (
+        dataset.take(train_size).shuffle(buffer_size=train_size).batch(batch_size)
+    )
+    test_dataset = dataset.skip(train_size).batch(batch_size)
+
+    return train_dataset, test_dataset
```
Comment (fchollet) [inline examples/generative/bayesian_neural_networks_wine.py:80]: Please add prefetching and caching. Please also add a comment stating that:

- We shuffle with a buffer the same size as the dataset because the dataset is very small and fits into memory.
- Same argument for caching.

File: examples/generative/bayesian_neural_networks_wine.py (hunk 3 above)
Comment (fchollet) [reply examples/generative/bayesian_neural_networks_wine.py:103]: I recommend adding a `validation_data` so we can monitor validation performance during training. Currently we only monitor *test* performance *after* training.

File: examples/generative/bayesian_neural_networks_wine.py (hunk 5 above)
Comment (fchollet) [reply examples/generative/bayesian_neural_networks_wine.py:149]: In this case I strongly recommend just using `Normalization`: feature-wise normalization with statistics computed beforehand is the standard to normalize inputs. BN only makes sense for normalizing intermediate activations.

Call adapt with a version of the dataset that yields the concatenated features. Note that if you do this you can also simplify the model by only having a single input (the concatenated features).

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

**Comment by fchollet** [comment]
On a related note, can you fix the following issues we previously missed in the neural decision forest example:

```
UserWarning: Input dict contained keys ['fnlwgt'] which did not match any model input. They will be ignored by the model.
```
([link](https://keras.io/examples/structured_data/deep_neural_decision_forests/#experiment-1-train-a-decision-tree-model))

We should simply drop the feature from the dataset when we create dataframes. Another minor note is that we should reserve a validation dataset to monitor performance during training.

File: examples/generative/bayesian_neural_networks_wine.py (hunk 6 above)
Comment (ksalama) [reply examples/generative/bayesian_neural_networks_wine.py:80]: Done

File: examples/generative/bayesian_neural_networks_wine.py (hunk 3 above)
Comment (ksalama) [reply examples/generative/bayesian_neural_networks_wine.py:103]: Added.

**Comment by ksalama** [comment]
@fchollet - Thanks for your comments. I have addressed them.

Will also update the neural decision forests one.

File: examples/generative/bayesian_neural_networks_wine.py (hunk 3 above)
Comment (fchollet) [reply examples/generative/bayesian_neural_networks_wine.py:108]: You cannot use `validation_split` with a dataset, it's only for indexable data (like a Numpy array or a dataframe). You should be seeing a warning or an error (if not, let me know, there should be one). You have to actually create a validation dataset beforehand and pass it as the `validation_data` argument.

File: examples/generative/bayesian_neural_networks_wine.py (hunk 5 above)
Comment (fchollet) [reply examples/generative/bayesian_neural_networks_wine.py:154]: Please address :)

**Comment by ksalama** [comment]
@fchollet I generated the .ipynb and .md files as thought the example is ready. Let me know if I need to do any further updates.

**Comment by fchollet** [comment]
Looks good, thank you! I've published it. Note that I've moved the example to the keras-recipes section, since it isn't generative deep learning, and it isn't really domain-specific -- you could apply the same techniques to almost any problem.
